File: core/src/crypto.cpp
#include "../include/crypto.h"
#include <algorithm>
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <stdexcept>

MemoryGuard::MemoryGuard(void* buffer, std::size_t size)
    : buffer_(buffer), size_(size) {}

MemoryGuard::~MemoryGuard() {
    if (buffer_ && size_ > 0) {
        OPENSSL_cleanse(buffer_, size_);
    }
}

std::string CryptoEngine::generate_salt() {
    constexpr size_t SALT_LEN = 16;
    std::string salt;
    salt.resize(SALT_LEN);

    // Заполняем буфер случайными байтами через OpenSSL RAND_bytes.
    if (RAND_bytes(reinterpret_cast<unsigned char*>(&salt[0]), SALT_LEN) != 1) {
        throw std::runtime_error("Failed to generate salt with OpenSSL RAND_bytes");
    }

    return salt;
}

std::string CryptoEngine::hash_password(const std::string& password, const std::string& salt) {
    constexpr int HASH_LEN = 32;
    constexpr int ITERATIONS = 100000;

    std::string hash;
    hash.resize(HASH_LEN);

    // Выполняем PBKDF2-HMAC-SHA256 с указанным количеством итераций.
    if (PKCS5_PBKDF2_HMAC(
            password.c_str(), static_cast<int>(password.size()),
            reinterpret_cast<const unsigned char*>(salt.data()), static_cast<int>(salt.size()),
            ITERATIONS,
            EVP_sha256(),
            HASH_LEN,
            reinterpret_cast<unsigned char*>(&hash[0])) != 1) {
        throw std::runtime_error("Failed to hash password with PBKDF2-HMAC-SHA256");
    }

    return hash;
}

bool CryptoEngine::secure_compare(const std::string& a, const std::string& b) {
    const size_t len = std::max(a.size(), b.size());
    unsigned char diff = static_cast<unsigned char>(a.size() ^ b.size());

    for (size_t i = 0; i < len; ++i) {
        unsigned char ca = i < a.size() ? static_cast<unsigned char>(a[i]) : 0;
        unsigned char cb = i < b.size() ? static_cast<unsigned char>(b[i]) : 0;
        diff |= ca ^ cb;
    }

    return diff == 0;
}

File: app/crypto.py
"""Криптографическое ядро для защищённого мессенджера.

Содержит функции для генерации соли, хеширования пароля через
PBKDF2-HMAC-SHA256 (100 000 итераций) и безопасного сравнения строк.
"""

import hashlib
import hmac
import os
import secrets


PBKDF2_ITERATIONS = 100_000
SALT_LEN = 16
HASH_LEN = 32
HASH_ALGO = "sha256"


def generate_salt() -> bytes:
    """Сгенерировать случайную соль длиной 16 байт."""
    return secrets.token_bytes(SALT_LEN)


def hash_password(password: bytes, salt: bytes) -> bytes:
    """Хешировать пароль с солью через PBKDF2-HMAC-SHA256."""
    return hashlib.pbkdf2_hmac(
        HASH_ALGO,
        bytes(password),
        bytes(salt),
        PBKDF2_ITERATIONS,
        dklen=HASH_LEN,
    )


def secure_compare(a: bytes, b: bytes) -> bool:
    """Сравнить две строки за постоянное время (защита от timing attack)."""
    return hmac.compare_digest(a, b)
